kotlinx-coroutines-core/common/test/CancelledParentAttachTest.kt

```diff
@@ -11,75 +11,105 @@ import kotlin.test.*
 class CancelledParentAttachTest : TestBase() {
 
     @Test
-    fun testAsync() = CoroutineStart.values().forEach(::testAsyncCancelledParent)
+    fun testAsync() = runTest {
+        CoroutineStart.values().forEach { testAsyncCancelledParent(it) }
+    }
 
-    private fun testAsyncCancelledParent(start: CoroutineStart) =
-        runTest({ it is CancellationException }) {
-            cancel()
-            expect(1)
-            val d = async<Int>(start = start) { 42 }
-            expect(2)
-            d.invokeOnCompletion {
-                finish(3)
-                reset()
+    private suspend fun testAsyncCancelledParent(start: CoroutineStart) {
+        try {
+            withContext(Job()) {
+                cancel()
+                expect(1)
+                val d = async<Int>(start = start) { 42 }
+                expect(2)
+                d.invokeOnCompletion {
+                    finish(3)
+                    reset()
+                }
             }
+            expectUnreached()
+        } catch (e: CancellationException) {
+            // Expected
         }
+    }
 
     @Test
-    fun testLaunch() = CoroutineStart.values().forEach(::testLaunchCancelledParent)
+    fun testLaunch() = runTest {
+        CoroutineStart.values().forEach { testLaunchCancelledParent(it) }
+    }
 
-    private fun testLaunchCancelledParent(start: CoroutineStart) =
-        runTest({ it is CancellationException }) {
-            cancel()
-            expect(1)
-            val d = launch(start = start) { }
-            expect(2)
-            d.invokeOnCompletion {
-                finish(3)
-                reset()
+    private suspend fun testLaunchCancelledParent(start: CoroutineStart) {
```

Comment:
I've actually tried a less intrusive solution with chaining multiple `runTest` into a single promise, but it turned out to be almost impossible due to `GlobalScope.*` and undefined order between resolving of `runTest` promise and invocation of `invokeOnCompletion` in children bound to the resulting promise.